packages/design-system/src/components/IdleTimeout/IdleTimeout.tsx

```diff
@@ -77,70 +78,148 @@ const defaultMessageFormatter = (timeTilTimeout: number): React.ReactNode => {
   );
 };
 
+// names for local storage variables
+const timeoutCookieName = 'CMS_DS_TIMEOUT';
+const timeoutWarningCookieName = 'CMS_DS_WARNING';
+
 const IdleTimeout = ({
-  closeDialogText = 'Close',
   continueSessionText = 'Continue session',
   heading = 'Are you still there?',
   endSessionButtonText = 'Logout',
+  endSessionRedirectUrl = '/logout',
   formatMessage = defaultMessageFormatter,
-  onClose,
   onSessionContinue,
   onSessionForcedEnd,
   onTimeout,
   showSessionEndButton = false,
   timeToTimeout,
   timeToWarning = 5,
-}: IdleTimeoutProps) => {
-  const handleContinueSession = () => {
-    // reset countdown timer
-    if (onSessionContinue) {
-      onSessionContinue();
+}: IdleTimeoutProps): JSX.Element => {
+  if (timeToWarning > timeToTimeout) {
+    console.error(
+      'Error in TimeoutManager component. `timeToWarning` is greater or equal to `timeToTimeout`'
+    );
+  }
+  // convert minutes to milliseconds
+  const msToTimeout = timeToTimeout * 60000;
+  const msToWarning = (timeToTimeout - timeToWarning) * 60000;
+  const [checkStatusTime, setCheckStatusTime] = useState<number>(null);
+  const [warningIntervalId, setWarningIntervalId] = useState<NodeJS.Timeout>(null);
+  const [showWarning, setShowWarning] = useState<boolean>(false);
+  const [timeInWarning, setTimeInWarning] = useState<number>(timeToWarning);
+
+  // cleanup timeouts & intervals
+  const clearTimeouts = () => {
+    setCheckStatusTime(null);
+    if (warningIntervalId) {
+      clearInterval(warningIntervalId);
     }
   };
 
-  const handleEndSession = () => {
-    if (onSessionForcedEnd) {
-      // TODO: figure out if any params should be passed back to the app
-      onSessionForcedEnd();
-    } else if (onTimeout) {
-      onTimeout();
+  // when the countdown for the session ends, clean up, call callback & close modal
+  const handleTimeout = () => {
+    clearTimeouts();
+    removeEventListeners();
+    onTimeout();
+    setShowWarning(false);
+  };
+
+  // when it's time to warn the user about idleness,
+  // set an interval that updates the modal message
+  const handleWarningTimeout = () => {
+    setShowWarning(true);
+    removeEventListeners();
+    let timeTilTimeout = timeToWarning - 1;
+    const intervalId = setInterval(() => {
+      setTimeInWarning(timeTilTimeout);
+      timeTilTimeout--;
+    }, 60000);
```

Comment:
Could we simplify by actually using the same interval as the `checkWarningStatus` function? Like in that function can we `setTimeInWarning` based on subtracting the `warningTime` in that function from the current time?

Reply:
I think then the `timeInWarning` could be derrived from the current time and the state we already track.